fix: make `terragrunt-fetch-dependency-output-from-state` to work with not applied dependencies

rodrigorfk opened the pull request:
<!-- Prepend '[WIP]' to the title if this PR is still a work-in-progress. Remove it when it is ready for review! -->

## Description

Fixes #3349 and #2261.

If you enable the `--terragrunt-fetch-dependency-output-from-state` feature flag and runs a `terragrunt init` on a module that depends on a dependency that was not yet applied, terragrunt will fail, even if you are using `mock_outputs`. This fix is treating the returned error and checking if a `NoSuchKey` error was produced, allowing terragrunt to proceed if that was the case.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

<!-- One-line description of the PR that can be included in the final release notes. -->
Fix `--terragrunt-fetch-dependency-output-from-state` feature flag and allows it to work with not applied dependencies

### Migration Guide

<!-- Important: If you made any backward incompatible changes, then you must write a migration guide! -->

denis256 commented:
Looks like `TestTerragruntMockOutputsFromRemoteState` is failing
```
=== RUN   TestTerragruntMockOutputsFromRemoteState
    integration_test.go:4513: Copying fixture-output-from-remote-state to /tmp/terragrunt-test4260963045
    integration_test.go:4467: [terragrunt apply --terragrunt-fetch-dependency-output-from-state --auto-approve --terragrunt-non-interactive --terragrunt-working-dir /tmp/terragrunt-test4260963045/fixture-output-from-remote-state/env1/app1]

[0m[1mInitializing the backend...[0m
[0m[32m
Successfully configured the backend "s3"! OpenTofu will automatically
use this backend unless the backend configuration changes.[0m

[0m[1mInitializing provider plugins...[0m

[0m[1m[32mOpenTofu has been successfully initialized![0m[32m[0m
[0m[32m
You may now begin working with OpenTofu. Try running "tofu plan" to see
any changes that are required for your infrastructure. All OpenTofu commands
should now work.

If you ever set or change modules or backend configuration for OpenTofu,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.[0m

Changes to Outputs:
  [32m+[0m[0m app1_text = "app1 output"

You can apply this plan to save these new output values to the OpenTofu
state, without changing any real infrastructure.
[0m[1m[32m
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
[0m[0m[1m[32m
Outputs:

[0mapp1_text = "app1 output"
    integration_test.go:4467: [terragrunt init --terragrunt-fetch-dependency-output-from-state --terragrunt-non-interactive --terragrunt-working-dir /tmp/terragrunt-test4260963045/fixture-output-from-remote-state/env1/app2]
    integration_test.go:5371: 
        	Error Trace:	/home/circleci/project/test/integration_test.go:5371
        	Error:      	Should be true
        	Test:       	TestTerragruntMockOutputsFromRemoteState
    integration_test.go:4824: Deleting test s3 bucket terragrunt-test-bucket-eseh5d
--- FAIL: TestTerragruntMockOutputsFromRemoteState (5.92s)
```

rodrigorfk commented:
Thanks for the update @denis256 , I added a check that is terraform only and therefore it fails with tofu, I've amended the PR and fixed it, could you please run the tests once again? thanks in advance.

denis256 commented on test/integration_test.go:
```diff
@@ -5337,6 +5337,40 @@ func TestTerragruntOutputFromRemoteState(t *testing.T) { //nolint: paralleltest
+	rootTerragruntConfigPath := util.JoinPath(tmpEnvPath, TEST_FIXTURE_OUTPUT_FROM_REMOTE_STATE, config.DefaultTerragruntConfigPath)
+	copyTerragruntConfigAndFillPlaceholders(t, rootTerragruntConfigPath, rootTerragruntConfigPath, s3BucketName, "not-used", "not-used")
+
+	environmentPath := fmt.Sprintf("%s/%s/env1", tmpEnvPath, TEST_FIXTURE_OUTPUT_FROM_REMOTE_STATE)
```
I think `filepath.Join()` will be more cross-platform

rodrigorfk replied:
I actually copied that code from the `TestTerragruntOutputFromRemoteState` test above, but sure, make sense and thanks for the suggestion, I've fixed it.

denis256 commented on test/integration_test.go:
```diff
@@ -5337,6 +5337,40 @@ func TestTerragruntOutputFromRemoteState(t *testing.T) { //nolint: paralleltest
+		stdout bytes.Buffer
+		stderr bytes.Buffer
+	)
+	runTerragruntRedirectOutput(t, fmt.Sprintf("terragrunt init --terragrunt-fetch-dependency-output-from-state --terragrunt-non-interactive --terragrunt-working-dir %s/app2", environmentPath), &stdout, &stderr)
```
`runTerragruntCommandWithOutput()` - will also print output to terminal, will be helpful to debug why it is crashing

rodrigorfk replied:
thanks for the suggestion, I've updated it.
